src/main/java/com/xceptance/xlt/engine/htmlunit/okhttp3/OkHttp3WebConnection.java

```diff
@@ -0,0 +1,380 @@
+import com.xceptance.xlt.engine.htmlunit.AbstractWebConnection;
+
+import okhttp3.Authenticator;
+import okhttp3.ConnectionPool;
+import okhttp3.Dispatcher;
+import okhttp3.Headers;
+import okhttp3.MediaType;
+import okhttp3.MultipartBody;
+import okhttp3.OkHttpClient;
+import okhttp3.OkHttpClient.Builder;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+/**
+ * An alternative {@link WebConnection} implementation that uses OkHttp3 as HTTP client.
+ */
+public class OkHttp3WebConnection extends AbstractWebConnection<OkHttpClient, Request, Response>
+{
+    // TODO: stale connection check?
+    // TODO: tcp nodelay?
+
+    private static final List<Protocol> HTTP_1_1_ONLY = Arrays.asList(Protocol.HTTP_1_1);
+
+    private static final List<Protocol> HTTP_2_AND_1_1 = Arrays.asList(Protocol.HTTP_2, Protocol.HTTP_1_1);
+
+    private static final EasyHostnameVerifier INSECURE_HOSTNAME_VERIFIER = new EasyHostnameVerifier();
+
+    private static final EasyX509TrustManager INSECURE_TRUST_MANAGER = new EasyX509TrustManager(null);
+
+    private static final SSLSocketFactory INSECURE_SSL_SOCKET_FACTORY = createInsecureSslSocketFactory();
+
+    private final AuthenticationCache authenticationCache;
+
+    private final ConnectionPool connectionPool;
+
+    private final DnsImpl dns;
+
+    private final List<Protocol> protocols;
+
+    /**
+     * Constructor.
+     *
+     * @param webClient
+     *            the owning web client
+     * @param http2Enabled
+     *            whether or not HTTP/2 is enabled at all
+     */
+    public OkHttp3WebConnection(final WebClient webClient, final boolean http2Enabled)
+    {
+        super(webClient);
+
+        authenticationCache = new AuthenticationCache();
+        connectionPool = new ConnectionPool(6, 60, TimeUnit.SECONDS);
+        dns = new DnsImpl(new XltDnsResolver());
+        protocols = http2Enabled ? HTTP_2_AND_1_1 : HTTP_1_1_ONLY;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected OkHttpClient createHttpClient(final WebClient webClient, final WebRequest webRequest) throws Exception
+    {
+        final WebClientOptions webClientOptions = webClient.getOptions();
+        final Authenticator authenticator = new AuthenticatorImpl(webClient.getCredentialsProvider());
+
+        final Builder httpClientBuilder = new OkHttpClient.Builder();
+
+        // basic settings and components
+        httpClientBuilder.authenticator(authenticator);
+        // httpClientBuilder.cache(null);
+        httpClientBuilder.connectionPool(connectionPool);
+        // httpClientBuilder.connectionSpecs(createConnectionSpecs(webClientOptions));
+        httpClientBuilder.cookieJar(new CookieJarImpl(webClient.getCookieManager()));
+        httpClientBuilder.dns(dns);
+        httpClientBuilder.followRedirects(false);
+        httpClientBuilder.followSslRedirects(false);
```

Comment:
Maybe we should add a comment why we disable automatic following of redirects.

Reply:
Done.